java/src/jmri/jmrit/display/GlobalVariableIcon.java

```diff
@@ -0,0 +1,507 @@
+        return namedGlobalVariable;
+    }
+
+    public GlobalVariable getGlobalVariable() {
+        if (namedGlobalVariable == null) {
+            return null;
+        }
+        return namedGlobalVariable.getBean();
+    }
+
+    @Override
+    public jmri.NamedBean getNamedBean() {
+        return getGlobalVariable();
+    }
+
+    public java.util.HashMap<String, NamedIcon> getMap() {
+        return map;
+    }
+
+    // display icons
+    public void addKeyAndIcon(NamedIcon icon, String keyValue) {
+        if (map == null) {
+            setMap(); // initialize if needed
+        }
+        map.put(keyValue, icon);
+        // drop size cache
+        //height = -1;
+        //width = -1;
+        displayState(); // in case changed
+    }
+
+    // update icon as state of GlobalVariable changes
+    @Override
+    public void propertyChange(java.beans.PropertyChangeEvent e) {
+        if (log.isDebugEnabled()) {
+            log.debug("property change: {} is now {}",
+                    e.getPropertyName(), e.getNewValue());
+        }
+        if (e.getPropertyName().equals("value")) {
+            displayState();
+        }
+        if (e.getSource() instanceof jmri.Throttle) {
+            if (e.getPropertyName().equals(jmri.Throttle.ISFORWARD)) {
+                Boolean boo = (Boolean) e.getNewValue();
+                if (boo) {
+                    flipIcon(NamedIcon.NOFLIP);
+                } else {
+                    flipIcon(NamedIcon.HORIZONTALFLIP);
+                }
+            }
+        }
+    }
+
+    @Override
+    public String getNameString() {
+        String name;
+        if (namedGlobalVariable == null) {
+            name = Bundle.getMessage("NotConnected");
+        } else {
+            name = getGlobalVariable().getDisplayName(DisplayOptions.USERNAME_SYSTEMNAME);
+        }
+        return name;
+    }
+
+    public void setSelectable(boolean b) {
+        selectable = b;
+    }
+
+    public boolean isSelectable() {
+        return selectable;
+    }
+    boolean selectable = false;
+
+    @Override
+    public boolean showPopUp(JPopupMenu popup) {
+        if (isEditable() && selectable) {
+            popup.add(new JSeparator());
+
+            for (String key : map.keySet()) {
+                //String value = ((NamedIcon)map.get(key)).getName();
+                popup.add(new AbstractAction(key) {
+
+                    @Override
+                    public void actionPerformed(ActionEvent e) {
+                        String key = e.getActionCommand();
+                        setValue(key);
+                    }
+                });
+            }
+            return true;
+        }  // end of selectable
+        return false;
+    }
+
+    /**
+     * Text edits cannot be done to GlobalVariable text - override
+     */
+    @Override
+    public boolean setTextEditMenu(JPopupMenu popup) {
+        popup.add(new AbstractAction(Bundle.getMessage("EditGlobalVariableValue")) {
```

Comment:
java.util.MissingResourceException: Resource 'EditGlobalVariableValue' not found

Reply:
The MissingResourceException is fixed.

How do you trigger the error:
`java.lang.ClassCastException: class jmri.jmrit.display.GlobalVariableSpinnerIcon cannot be cast to class jmri.jmrit.display.PositionableLabel` ?

I'm failing to reproduce this error. Which steps do you take? Does it work for a MemorySpinnerIcon?

Reply:
1. Add a global variable icon to the a **Panel Editor** panel as a spinner or input box.
2. Do a right click and select Properties
3. Click on **Apply**